### PR TITLE
feat(osm-config): osm config go types and client

### LINKED ISSUE #3037: config: Generate Go Client code for new OSM Config CRD
This issue is for the generation of the Go client code for the newly created OSM Configuration CRD (#3033)

With #3035 we created a bash script helper, which generates the code based on the YAML created with #3033 and #3034. This issue here is for the addition of the generated code to the OSM repo.

Example of generated code:  https://github.com/openservicemesh/osm/pull/933/files

(There is a lot of it - it would be great to commit it in one PR where the PR contains only the auto-generated code)

---

This issue is blocked / depends on:
  - [ ] config: Create code gen helper for generation of Go Client code for new OSM Config CRD #3035

---

All related issues:
  - [ ] config: Design a new OSM Config CRD (YAML only) #3033
  - [ ] config: Add validation to the new OSM Config CRD (in YAML) #3034
  - [ ] config: Create code gen helper for generation of Go Client code for new OSM Config CRD #3035
  - [ ] config: Generate Go Client code for new OSM Config CRD #3037
  - [ ] config: Create new Configuration CRD informers #3038
  - [ ] config: Swap ConfigMap usage with Configuration CRD in the pkg/configurator module #3039
  - [ ] config: Update documentation to reflect the usage of a Configuration CRD (not a ConfigMap); Cleanup #3040
  - [ ] config: Update ValidatingWebhookConfig and handler (for osm-config ConfigMap changes) #3041

### PR DESCRIPTION
**Pull request by allenlsy**
Signed-off-by: Allen Leigh <allenlsy@gmail.com>

**Description**:

Add OsmConfig v1alpha1 client-go style types. Generate clientset, informer and lister using `make codegen`.

This is part of the process of migrating from osm-configmap to OsmConfig CRD.

**Affected area**:

- New Functionality      [X]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

No

**Related issues**

close #3037

### CODE REVIEW

**Review by shashankram — COMMENTED** [review]
This looks good. Few things:
1. Ignore golangci-lint for autogenerated code: `.golangci.yml`
2. Ignore autogenerated code from code coverage reports: `scripts/test-w-coverage.sh`

**Comment by allenlsy** [comment]
@shashankram  Added `--skip-dirs pkg/gen` in main.yaml

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -44,7 +44,7 @@ jobs:
         uses: golangci/golangci-lint-action@v2
         with:
           version: v1.30
-          args: --timeout 5m
+          args: --timeout 5m --skip-dirs pkg/gen
```
Comment (shashankram) [inline .github/workflows/main.yml:47]: also pkg/apis?

Comment (allenlsy) [reply .github/workflows/main.yml:47]: sure

Comment (allenlsy) [reply .github/workflows/main.yml:47]: I think you are right. deepcopy code is generated. We can ignore that file.

File: scripts/test-w-coverage.sh [hunk 2]
```diff
@@ -19,7 +19,7 @@ go test -timeout 120s \
    -covermode count "${modules[@]}" | tee testoutput.txt || { echo "go test returned non-zero"; exit 1; }
 
 # shellcheck disable=SC2002
-cat coverage.txt.with_generated_code | grep -v "_generated.go" | grep -v "fake.go" > coverage.txt
+cat coverage.txt.with_generated_code | grep -v "_generated.go" | grep -v "fake.go" | grep -v "zz_generated.deepcopy.go" > coverage.txt
```
Comment (shashankram) [inline scripts/test-w-coverage.sh:22]: Let's skip using `grep -v pkg/gen | grep -v pkg/apis`

File: codegen/gen-crd-client.sh [hunk 3]
```diff
@@ -71,5 +71,5 @@ function generate_client() {
   rm -rf ${TEMP_DIR}
 }
 
-echo "##### Generating config.openservicemesh.io client ######"
-generate_client "config" "v1alpha1"
\ No newline at end of file
+echo "##### Generating osmconfig.openservicemesh.io client ######"
+generate_client "osmconfig" "v1alpha1"
```
Comment (shashankram) [inline codegen/gen-crd-client.sh:75]: This should be `config` because that's the group prefix for the group `config.openrservicemesh.io`

**Review by shashankram — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -44,7 +44,7 @@ jobs:
         uses: golangci/golangci-lint-action@v2
         with:
           version: v1.30
-          args: --timeout 5m
+          args: --timeout 5m --skip-dirs pkg/gen --skip-files "zz_generated.deepcopy.go$"
```
Comment (draychev) [inline .github/workflows/main.yml:47]: Is the `$` at the intentional?

**Review by draychev — APPROVED** [review]
This is looking great! 
Thank you @allenlsy 
Thank you @shashankram

File: .github/workflows/main.yml (hunk 4 above)
Comment (allenlsy) [reply .github/workflows/main.yml:47]: Yes, `--skip-files` takes a regex. [Reference here](https://golangci-lint.run/usage/configuration/):

```
  # which files to skip: they will be analyzed, but issues from them
  # won't be reported. Default value is empty list, but there is
  # no need to include all autogenerated files, we confidently recognize
  # autogenerated files. If it's not please let us know.
  # "/" will be replaced by current OS file path separator to properly work
  # on Windows.
  skip-files:
    - ".*\\.my\\.go$"
    - lib/bad.go
```

File: charts/osm/crds/meshconfig.yaml [hunk 5]
```diff
@@ -17,17 +17,17 @@
 apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
-  name: osmconfigs.config.openservicemesh.io
+  name: meshes.config.openservicemesh.io
```
Comment (shashankram) [inline charts/osm/crds/meshconfig.yaml:20]:
```suggestion
  name: meshconfigs.config.openservicemesh.io
```

File: charts/osm/crds/meshconfig.yaml [hunk 6]
```diff
@@ -17,17 +17,17 @@
...
-    listKind: OsmConfigList
+    kind: MeshConfig
+    listKind: MeshConfigList
     shortNames:
-      - osmconfig
-    plural: osmconfigs
-    singular: osmconfig
+      - meshcfg
+    singular: mesh
+    plural: meshes
```
Comment (shashankram) [inline charts/osm/crds/meshconfig.yaml:30]:
```suggestion
    plural: meshconfigs
```

File: charts/osm/crds/meshconfig.yaml [hunk 7]
```diff
@@ -17,17 +17,17 @@
...
-    kind: OsmConfig
-    listKind: OsmConfigList
+    kind: MeshConfig
+    listKind: MeshConfigList
     shortNames:
-      - osmconfig
-    plural: osmconfigs
-    singular: osmconfig
+      - meshcfg
+    singular: mesh
```
Comment (shashankram) [inline charts/osm/crds/meshconfig.yaml:29]:
```suggestion
    singular: meshconfig
```

File: pkg/apis/config/v1alpha1/mesh_config.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+// SidecarSpec is the spec for OSM's sidecar configuration
+type SidecarSpec struct {
+	EnablePrivilegedInitContainer bool   `json:"enable_privileged_init_container,omitempty"`
+	LogLevel                      string `json:"log_level,omitempty" default:"error"`
+}
+
+// TrafficSpec is the spec for OSM's traffic management configuration
+type TrafficSpec struct {
+	Egress                            bool   `json:"egress,omitempty"`
+	OutboundIPRangeExclusionList      string `json:"outbound_ip_range_exclusion_list,omitempty"`
```
Comment (shashankram) [inline pkg/apis/config/v1alpha1/mesh_config.go:33]: This should be a list of strings.
Please track this in a new issue and address in a subsequent PR.

**Review by shashankram — CHANGES_REQUESTED** [review]
Looks good overall.
Few blockers before this is good to go.

File: pkg/apis/config/v1alpha1/mesh_config.go (hunk 8 above)
Comment (allenlsy) [reply pkg/apis/config/v1alpha1/mesh_config.go]: Updated in this PR

**Review by shashankram — APPROVED** [review]
Looks great.

We will need to update the CRD to make `outbound_ip_range_exclusion_list` an array type before this code is usable within the controller.

**Pull request merged**